uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/PortletPrefsRESTController.java

```diff
@@ -0,0 +1,679 @@
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+
+/**
+ * PortletPrefsRESTController provides REST targets for preference operations for soffits and
+ * webcomponents.
+ */
+@Controller
+public class PortletPrefsRESTController {
+
+    protected final Log logger = LogFactory.getLog(getClass());
+    private ObjectMapper mapper;
+
+    private IPersonManager personManager;
+
+    private final boolean DEFINITION_MODE = true;
+    private final boolean ENTITY_MODE = false;
+
+    @Autowired
+    public void setPersonManager(IPersonManager personManager) {
+        this.personManager = personManager;
+    }
+
+    private IPortletDefinitionDao portletDao;
+
+    @Autowired
+    public void setPortletDao(IPortletDefinitionDao portletDao) {
+        this.portletDao = portletDao;
+    }
+
+    private IPortletEntityRegistry portletEntityRegistry;
+
+    @Autowired
+    public void setPortletEntityRegistry(IPortletEntityRegistry portletEntityRegistry) {
+        this.portletEntityRegistry = portletEntityRegistry;
+    }
+
+    private PortletPreferencesFactory portletPreferencesFactory;
+
+    @Autowired
+    @Qualifier("portletPreferencesFactory")
+    public void setPortletPreferencesFactoryAPI(
+            PortletPreferencesFactory portletPreferencesFactory) {
+        this.portletPreferencesFactory = portletPreferencesFactory;
+    }
+
+    public PortletPrefsRESTController() {
+        this.mapper = new ObjectMapper();
+    }
+
+    /**
+     * Get uPortal Entity information for portlet/soffit/webcomponent. This method provides a REST
+     * interface for obtaining uPortal portlet/soffit/webcomponent metadata
+     *
+     * <p>The path for this method is /prefs/getentity/fname. The fname is a string representing the
+     * fname of the portlet/soffit/webcomponent to find the entity information for
+     */
+    /*non-javadoc
+     * I used this primarily for testing, could be useful given that it returns the windowstate, etc
+     * up to you if you want to keep it or modify it
+     * */
+    @RequestMapping(value = "/prefs/getentity/{fname}", method = RequestMethod.GET)
+    public ResponseEntity getEntity(
+            HttpServletRequest request,
+            HttpServletResponse response,
+            @ApiParam(value = "The portlet fname to get entity for") @PathVariable(value = "fname")
+                    String fname) {
+        try {
+            IPortletDefinition portletDefinition = portletDao.getPortletDefinitionByFname(fname);
+            if (portletDefinition == null) {
+                return new ResponseEntity<>("ERROR: Portlet not found", HttpStatus.NOT_FOUND);
+            }
+            IPortletEntity entity =
+                    portletEntityRegistry.getOrCreateDefaultPortletEntity(
+                            request, portletDefinition.getPortletDefinitionId());
+            StringBuilder stringBuilder = new StringBuilder();
```

Comment:
I feel that the JSON response should be built using POJOs and converted to JSON using something such as Jackson. Is there a reason behind building the response by hand?

Reply:
you can take [this REST API as example](https://github.com/Jasig/uPortal/blob/master/uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/search/SearchRESTController.java)